### ISSUE TITLE
[millipede, photonics-service, photospline] Merge Spline-optimization branch (Trac #1925)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1917]
Merge this branch into the trunk of millipede.

http://code.icecube.wisc.edu/projects/icecube/log/IceCube/sandbox/pingusoft/branches/Spline-optimization

Migrated from https://code.icecube.wisc.edu/ticket/1925
```json
{
    "status": "closed", 
    "changetime": "2017-05-18T14:36:53", 
    "description": "Merge this branch into the trunk of millipede.\n\nhttp://code.icecube.wisc.edu/projects/icecube/log/IceCube/sandbox/pingusoft/branches/Spline-optimization", 
    "reporter": "olivas", 
    "cc": "losecco, olivas, kjmeagher, andrii.terliuk, jvansanten", 
    "resolution": "wontfix", 
    "_ts": "1495118213615203", 
    "component": "combo reconstruction", 
    "summary": "[millipede, photonics-service, photospline] Merge Spline-optimization branch", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2016-12-20T15:51:55", 
    "milestone": "Long-Term Future", 
    "owner": "jbraun", 
    "type": "task"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170112T18:38:17: 
* olivas changed summary from "[millipede] Merge Spline-optimization branch" to "[millipede, photonics-service, photospline] Merge Spline-optimization branch"
* olivas changed cc from "losecco, olivas, kjmeagher" to "losecco, olivas, kjmeagher, jvs, andrii.terliuk"

**Comment by icecube-trac** [comment]
Trac update at 20170112T18:38:57: olivas changed cc from "losecco, olivas, kjmeagher, jvs, andrii.terliuk" to "losecco, olivas, kjmeagher, andrii.terliuk, jvansanten"

**Comment by icecube-trac** [comment]
Trac update at 20170112T18:39:39: olivas commented:

> For more information: https://drive.google.com/file/d/0BzNJUKbKnSxNTGJSY01zMlpFUGM/view

**Comment by icecube-trac** [comment]
Trac update at 20170116T13:26:07: 
* jvansanten commented:

> Before starting work on merging [parts of] this into trunk, I'd like to know whether PINGU people are actually using this, and what the real-world experience has been like. In particular, I'm not sure I understand whether reordering the source/receiver loop should make a difference in general, or only for particular zenith angles of the source where moving along the hypothetical track very different from moving along the strings in reverse. It would also be nice to see a demonstration of the performance improvement with a deterministic minimizer (in particular, obtaining exactly the same result in the same number of calls, only faster).
> 
> Also, special-casing the dimensions of splines found in the wild is in principle a thing that is worth doing, but I'd like to have a more general implementation. In particular, Chris Weaver has a C++11 port of photospline that is templated on the orders of the splines in the tensor product that realizes the same potential performance improvements, for the same reasons. If we do want to employ such an optimization, it would be good to bring Chris' port in from the cold. We've discussed this a few times, but neither of us has had the free_time*urgency factor to merge the changes back.


* jvansanten changed milestone from "" to "Long-Term Future"

**Comment by icecube-trac** [comment]
Trac update at 20170502T16:22:41: kjmeagher commented:

> What is the status of this bug, Are the Low-en people still pushing for them, do they have answers to jvs' comments? will this branch have to be altered to deal with the new photospline library?
>

**Comment by icecube-trac** [comment]
Trac update at 20170502T16:31:05: jvansanten commented:

> The changes in this branch would have to be merged back by hand in any case, since it doesn't include history information.
> 
> Not much work is left to do, though, since the ideas behind John's changes to photospline are included in photospline 2 as part of a more general implementation of such optimizations.
> 
> If a Low-En volunteer can be found to provide a performance test case, we can look at reordering the source-receiver loop in millipede.

**Comment by icecube-trac** [comment]
Trac update at 20170518T14:36:53: 
* kjmeagher changed resolution from "" to "wontfix"
* kjmeagher commented:

> Since the most important optimization in this branch is already implemented in photospline 2.0 and no one in the Low Energy group is using this branch, I am marking this won't fix


* kjmeagher changed status from "new" to "closed"
